### PR TITLE
Fix All Compiler Warnings

### LINKED ISSUE #779: Fix All Compiler Warnings
### Description of the task
With our migration to bazel, we have started to get a _lot_ of compiler warnings that we just didn't have turned on before. We should fix them all and make warnings fail compilation.

### Acceptance criteria

- [ ] Make warnings errors (so they will fail compilation)
- [ ] Fix all warnings in the codebase

### PR DESCRIPTION
**Pull request by amrm3lm**
### Description
added warning as error compiler option and fixed all warnings that resulted. 

Because of the way we designed our primitives and intents, we had lots of "-Woverloaded-virtual" warnings for overloading virtual functions. Because there are many instances of this warning and no way to easily modify the code to fix it (to my knowledge), i ended up suppressing this specific warning. 

Some performance tests contained unused variables (ex cost_function_test.cpp ) , but thats only because those variables were used to print the test duration (which was commented out).  So I did not remove the variables, just separated declaration and assignment, which does not cause a warning. 

### Testing Done
everything compiles warning free
ran and passed all tests

### Resolved Issues
resolves #779 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by henryymliu** [comment]
I'll let others review first.

Although that being said, are you sure suppressing `-Woverloaded-virtual` is the best solution? Can you provide an example where this warning is triggered? AFAIK there are a few ways to fix that warning:
- if the overload is unintended, simply rename the offending function
- declaring a private overload of the base class function without a definition
- I think something with `using <base func>;` in the class declaration works as well

If these warnings are widespread, and the overload is intentional, and if it would add a ton of 'dead code' to fix those warnings, then maybe I can justify suppressing that warning.

**Comment by MathewMacDougall** [comment]
@scveloso @EvanMorcom please review

**Comment by amrm3lm** [comment]
> Although that being said, are you sure suppressing `-Woverloaded-virtual` is the best solution? Can you provide an example where this warning is triggered?  AFAIK there are a few ways to fix that warning:
> 
>     * if the overload is unintended, simply rename the offending function
> 
>     * declaring a private overload of the base class function without a definition
> 
>     * I think something with `using <base func>;` in the class declaration works as well
> 
> 
> If these warnings are widespread, and the overload is intentional, and if it would add a ton of 'dead code' to fix those warnings, then maybe I can justify suppressing that warning.

I believe all the warnings are caused by the `accept()` function, which is defined as a virtual function in both `Intent.h` and `Primitive.h`. All intents cause that warning. For example in `stop_intent.h`, the class inherits `StopPrimitive` (which inherits `Primitive`) and `Intent`. My current understanding is that  when `stop_intent` uses `accept()`, it overloads `Intent::accept` and hides `Primitive:accept`. 

Adding `using Intent::accept;` to `stop_intent.h` does not fix the warning, however adding `using Primitive::accept`, bringing it into scope, does, so my guess is that its complaining about hiding functions and not overloading them. 

So i could add `using Primitive::accept` to all the `*_intent.h` files, but I felt it makes the code a bit confusing since we dont actually use `Primitive::accept`. 

Changing function names is also possible, although then Id have to make a lot of changes.  

>     * declaring a private overload of the base class function without a definition

Also I dont really understand what this would look like, can you please clarify?

**Comment by garethellis0** [comment]
I think that's _almost_ correct. An important note here is that while `StopIntent` overrides `void accept(IntentVisitor& visitor)`, it does not hide `void accept(PrimitiveVisitor& visitor)` (depending on your definition of "hide"). So depending on what visitor type (`IntentVisitor` or `PrimitiveVisitor`) is pased to `accept`, a different version of the function will be called. (So it's not hiding `accept(PrimitiveVisitor)` in the sense that it can still be called)

In terms of the fixes that you've suggested:
- I agree that adding `using Primitive::accept` to all `*_intent.h` files is probably not a great solution. I think it would confuse people more then anything else
- I am strongly against renaming, just because `accept` is a function that is part of the Visitor design pattern, and people have a certain expectation about functions being named a certain way for that pattern. Changing it would also induce confusion.

Currently I still think it's reasonable to disable this warning.

**Comment by EvanMorcom** [comment]
I think it makes the most sense to disable the warning as a quick fix. We get 95% of the benefit without having to spend a whole bunch more time on this. 

IMO, merge as-is and create a ticket to fix the compiler warning hack. 

We should focus on more important things RN

File: src/software/ai/passing/cost_function_test.cpp
Comment (EvanMorcom) [inline src/software/ai/passing/cost_function_test.cpp]: Nit:

Do we care about using a static_cast here?

**Review by EvanMorcom — DISMISSED** [review]
LGTM :100:

Got one nit comment, and pending what we plan to do with the compiler warning

**Review by scveloso — DISMISSED** [review]
LGTM, pending other remaining comments

File: src/software/ai/passing/cost_function_test.cpp
Comment (MathewMacDougall) [reply src/software/ai/passing/cost_function_test.cpp]: Yes we should. Our style guide prefers `static_cast` in all possible cases.

**Review by MathewMacDougall — COMMENTED** [review]
LGTM pending evan's castin comment, and resolving the merge conflict (which I just created, sorry :sweat_smile: )

**Comment by garethellis0** [comment]
@EvanMorcom - I don't even think it's worth creating a ticket to remove the blacklisted warning flag. The added complexity (as detailed in the comments above) is not worth it.

File: src/software/ai/hl/stp/tactic/shoot_goal_tactic.h
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/shoot_goal_tactic.h]: what does this mean?

File: src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: just curious, what warning does this fix?

File: src/software/ai/primitive/direct_wheels_primitive_test.cpp
Comment (jonathanlew) [inline src/software/ai/primitive/direct_wheels_primitive_test.cpp]: `DirectWheelsPrimitive` takes `int16_t` for the argument for `mismatched_val`, so it's implicitly casting to `int16_t`. Consider fixing this test so we're not passing in a `double` **or** (if you think it's out of scope) at least static_cast it to `int16_t` and file an issue

File: src/software/ai/primitive/stop_primitive.h
Comment (jonathanlew) [inline src/software/ai/primitive/stop_primitive.h]: I thought we decided to ignore the warning that requires this

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good, just got a few questions

File: src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp
Comment (MathewMacDougall) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: Looks like it's unused. I'd prefer to just delete it in that case.

File: src/software/ai/hl/stp/tactic/shoot_goal_tactic.h
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/tactic/shoot_goal_tactic.h]: Looks like it's unused. I'd prefer to just delete it in that case.

File: src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp
Comment (MathewMacDougall) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: I saw something similar with Hardik, I think it's to do with implicit `const`-ness of the variable? If it is, does adding the `const` modifier fix it?

File: src/software/ai/primitive/direct_wheels_primitive_test.cpp
Comment (henryymliu) [reply src/software/ai/primitive/direct_wheels_primitive_test.cpp]: In this case I think declaring a variable for this test is overkill anyways, I prefer the status quo.

File: src/.bazelrc
Comment (henryymliu) [inline src/.bazelrc]: Is this intentional?

**Review by henryymliu — COMMENTED** [review]
Thanks for explaining the need to suppress that warning, that's fine with me.

File: src/.bazelrc
Comment (MathewMacDougall) [reply src/.bazelrc]: The remote cache actually got removed in #1062 so this will be gone entirely once conflicts are resolved.

File: src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp
Comment (amrm3lm) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: According to the comments in this specific test `PassingEvaluationTest`, 
>This test does not assert anything. Rather, It can be used to gauge how
    fast ratePass is running, and can be profiled in order to find areas
    of improvement for ratePass

Its just that `avg` is unused when the speed results are not printed (they are commented out). We could comment this out as well, but that would cause `duration` and `start_time` to be unused earlier in the test, causing the same warning. 

We could comment out all the lines that actually measure performance, but I thought that having multiple commented out lines in one test is confusing. 

Alternatively, we could remove all the lines that relate to measuring performance, but then this test serves no purpose.

File: src/software/ai/hl/stp/tactic/shoot_goal_tactic.h
Comment (amrm3lm) [reply src/software/ai/hl/stp/tactic/shoot_goal_tactic.h]: yeah ill remove

File: src/software/ai/primitive/direct_wheels_primitive_test.cpp
Comment (amrm3lm) [reply src/software/ai/primitive/direct_wheels_primitive_test.cpp]: I think my reasoning was that the test was called `test_inequality_operator_with_mismatched_back_right_wheel_power`. And by mismatched i thought type mismatch. But since its being implicitly casted,  i'll change the parameter to -100.

File: src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp
Comment (MathewMacDougall) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: Maybe we can disable the test and leave it all uncommented? Then if we want to profile we can enable it.

https://stackoverflow.com/questions/7208070/googletest-how-to-skip-a-test

Also @garethellis0 thoughts since this was your test?

Comment (garethellis0) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: Honestly I'm not a _huge_ fan of this test, it was written before competition out of necessity. I think the `GTEST_SKIP` macro is appropriate here.

Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: I agree

Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: We should document tests that we are disabling in a central doc @MathewMacDougall @garethellis0 suggestions?

Comment (garethellis0) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: Hmm, don't think that's maintainable. I don't see any issue with having a policy of never skipping tests that are actually doing validation work.

Comment (MathewMacDougall) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: Yea IMO 99.99999999% of tests should be run, and if one needs to be disabled there should be a ticket to fix whatever needs to be fixed to re-enable it. Agree a doc isn't maintainable.

**Comment by garethellis0** [comment]
@amrm3lm - please feel free to re-requiest review from people whenever you're ready for another round (see 7 here: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/requesting-a-pull-request-review)

**Comment by garethellis0** [comment]
Looks like just need to run formatting!

**Review by garethellis0 — DISMISSED** [review]

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1: :shipit:

**Review by jonathanlew — DISMISSED** [review]
LGTM! :+1:

**Review by garethellis0 — APPROVED** [review]
Just as a note for anyone wondering, we're not currently building the old firmware in CI, so that's why there are no changes there. In order to avoid a ton of conflicts, I think this should be merged first, then I can fix all the firmware issues after my big PR with the control changes is merged.

LGTM :+1: :tada: - very happy to see this going in

**Review by MathewMacDougall — APPROVED** [review]
:+1: :tada: thanks for pushing to get this done!

**Pull request merged**
